### DISCUSSION TITLE
Unit Menu Skills - How to add?

### DISCUSSION THREAD

**Discussion by JesterWizard** [MokhaLeee/fe8u-cskillsys-kernel#225] [General]
Is the skill system at a stage where unit menu skills can be added? I tried recently and got stuck around ``UnitMenu.c`` trying to get my text strings to display.

**Comment by MokhaLeee** [comment]
same as decomp

**Reply by JesterWizard** [reply]
So I got most of the way there. The effect works, but the text doesn't display?

![image](https://github.com/user-attachments/assets/f0484a50-a8cc-4e6e-8db0-9eb8483b75ee)
![image](https://github.com/user-attachments/assets/4f6a0c02-49e5-4e43-9b15-f1e68ad523b9)
![image](https://github.com/user-attachments/assets/480f8c23-3d8a-4aff-98c8-f13e0467a60d)
![image](https://github.com/user-attachments/assets/8af652ad-e904-44f5-bfe7-733e8cf2bb8d)
![image](https://github.com/user-attachments/assets/877e031a-40a4-4829-8c48-a8cc753e68e7)

**Reply by JesterWizard** [reply]
Ignore me. I know why

**Comment by MokhaLeee** [comment]
According to the text you mentioned, I think it better to put them as a combat-art skills. Maybe you can pull the skill info first and wait for me to give you a sample how to achieve skill effects via combat-art.

Menu related operation means introduing a lot of new code and form a very deep call stack. This will create a lot of corner cases, since they will cooperate with the battle system, it will be more difficult to repair if something goes wrong. At the same time, we also need to consider how to easily introduce more skills on this basis. All this requires rigorous code reviews with caution.

**Reply by JesterWizard** [reply]
I got it working finally. I'm sure you already know most of what you want to do, but maybe you can use what I did as a basis as you form a more robust system? I ended up placing the skill unit menu names and descriptions in `Texts.txt` and the skill usability and effect definitions in `skill-system.h`. I also placed the effect and usability instances in `MenuSkills.c`.

![image](https://github.com/user-attachments/assets/0c14be22-018c-4a96-b4c5-d158bf533d79)

What I have noticed is that the way `UnitMenu.c` is set up, the unit menu option name and descriptions require hardcoded addresses (take attack for example, `attack`, which uses `0x67B and 0x6C0`). Will this not have an impact if certain skills are removed or moved around? I think the names should be used instead.

**Comment by MokhaLeee** [comment]
Two issue you need to ansower:
1. If the player close the game during battle-anim, how can their pre-selected skills continue to take effect after next time they reopen the game
2. If player canceled during target-selection and switch to control another character, how can you ensure that the skill can truely avoid to still   take effect.

not to say you need to answer how to introduce more and more and more familiar skills.

**Reply by JesterWizard** [reply]
I haven't even gotten to that stage yet. Apparently I jumped the gun, as the skill doesn't take effect in battle. And to add a third point to your questions, I still need a way to make the skill take effect again after backing out of the battle forecast, returning to the weapon select screen and then navigating back to the battle forecast. I was setting a global flag for the skill in `GambleCommandEffect` here and then checking in the prebattle file for its value, but it doesn't persist in the above scenario. 

![image](https://github.com/user-attachments/assets/ba895eb9-03dd-4a57-ad7a-e63ee78e82c1)
![image](https://github.com/user-attachments/assets/c4c9c868-9996-45f6-b20b-ef1f705b2e4c)

**Comment by MokhaLeee** [comment]
https://github.com/MokhaLeee/fe8u-cskillsys-kernel/pull/227
